main/com.tunein/device.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+const Tp = require('thingpedia');
+const axios = require('axios');
+const axiosExtensions = require('axios-extensions');
+const url = require('url');
+
+const BASE_URL = 'http://opml.radiotime.com';
+const CONTENT_TYPE = 'application/x-www-form-urlencoded';
+const HTTP_STATUS_OK = 200;
+const SEARCH_PARAM = '/Search.ashx';
+const BROWSE_PARAM = '/Browse.ashx';
+const BROWSE_ITEMS = ['trending', 'music', 'sports', 'talk'];
+
+module.exports = class TuneinRadioDevice extends Tp.BaseDevice {
+    constructor(engine, state, args={}) {
+         super(engine, state);
+         this.uniqueId = 'com.tunein';
+         this.name = 'Tunein';
+         this.description = 'Stream radio through TuneIn API';
+         axios.defaults.headers.post['Content-Type'] = args.content_type || CONTENT_TYPE;
+         axios.defaults.baseURL = args.protocol || BASE_URL;
+         axios.defaults.params = { render: 'json' };
+    }
+
+    _get_tunein_response(req) {
+        return new Promise( (resolve, reject) => {
+            axios.get(req.url, req)
+                .then(response => { 
+                    if (response.data.head.status == HTTP_STATUS_OK) {
+                        let content = response.data.body;
+                        if ('c' in req.params && BROWSE_ITEMS.includes(req.params.c.toLowerCase())) {
+                            return resolve(content);
+                        } else if (content.find(item => item.text.toLowerCase() == 'stations')) {
+                            return resolve(
+                                content.filter(item => item.text.toLowerCase() == 'stations')
+                                       .flatMap(({children}) => children)
+                                       .filter(channel => (channel.type.toLowerCase() == 'audio' && channel.item.toLowerCase() == 'station'))
+                            );
+                        } else {
+                            return resolve(content.filter(channel => (channel.type.toLowerCase() == 'audio' && channel.item.toLowerCase() == 'station')));
+                        }; 
+                    } else {
+                        reject(response.data);
+                        return new Error(`HTTP request error: ${response.data.head.fault}`);
+                    }
+                })
+                .catch(err => throwError("streaming_service_unaccessible"));
+        }).catch(err => throwError("streaming_service_unaccessible"));
```

Comment:
You'll want to check what error this is. If it's a network error, let it bubble up, because Genie already knows how to handle those. If it's a programming error (TypeError or similar), also let it bubble up, so we show a generic error message and the full error is logged with the stack.